# Pacakges used
```{r}
library((Seurat)
library(dplyr)
library(xlsx)
```

# Subclustering
```{r}
uRA2.3 = subset(uRA2, subset = seurat_clusters == "3") # Isolate cells in cluster 3

# Single-Cell Transform
uRA2.3 = SCTransform(uRA2.3, variable.features.n = dim(uRA2.3)[1], assay = "RNA")

# Principal component analysis
uRA2.3 = RunPCA(uRA2.3, features = VariableFeatures(uRA2.3))
# ElbowPlot(uRA2.3, reduction = "pca", ndims = 50)

# Unsupervised re-clustering
uRA2.3 = FindNeighbors(uRA2.3, dims = 1:40)
uRA2.3 = FindClusters(uRA2.3, resolution = 0.95)

# Running UMAP
uRA2.3 = RunUMAP(uRA2.3, dims = 1:40)
# UMAPPlot(uRA2.3, label = TRUE)

# Rename clusters
Idents(uRA2.3, cells = c3.0 = WhichCells(uRA2.3, idents = "0")) = "3.0"
Idents(uRA2.3, cells = c3.0 = WhichCells(uRA2.3, idents = "1")) = "3.1"
Idents(uRA2.3, cells = c3.0 = WhichCells(uRA2.3, idents = "2")) = "3.2"

# Differential gene expression
uRA2.3.marks = subset(FindAllMarkers(uRA2.3, only.pos = FALSE, min.pct = 0.25), 
	subset = p_val_adj < 0.05)

pos.marks = subset(uRA2.3.marks, subset = avg_log2FC > 0.0)
neg.marks = subset(uRA2.3.marks, subset = avg_log2FC < 0.0)

write.xlsx(pos.marks %>% group_by(cluster) %>% top_n(-10, p_val_adj), 
	"reports/uRA2_C3_subs.xlsx", sheetName = "Top10_Up", col.names = TRUE, row.names = FALSE)
write.xlsx(neg.marks %>% group_by(cluster) %>% top_n(-10, p_val_adj), )
	"reports/uRA2_C3_subs.xlsx", sheetName = "Top10_Down", col.names = TRUE, row.names = FALSE, append = TRUE)

saveRDS(uRA2.3, file = "data/uRA2_C3_subs") # Save sub-cluster object


# Map subclusters to base data
uRA2.1 = uRA2
Idents(uRA2.1, cells = WhichCells(uRA2.3, idents = "0")) = "3.0"
Idents(uRA2.1, cells = WhichCells(uRA2.3, idents = "1")) = "3.1"
Idents(uRA2.1, cells = WhichCells(uRA2.3, idents = "2")) = "3.2"
Idents(uRA2.1) = factor(Idents(uRA2.1), 
	levels = c("0", "1", "2", "3.0", "3.1", "3.2","4", "5")
```
