# Packages used
```{r}
library((Seurat)
library(dplyr)
library(xlsx)
```

# Subclustering
```{r}
uRA4.3 = subset(uRA4, subset = seurat_clusters == "3") # Isolate cells in cluster 3

# Single-Cell Transform
uRA4.3 = SCTransform(uRA4.3, variable.features.n = dim(uRA4.3)[1], assay = "RNA")

# Principal Component Analysis
uRA4.3 = RunPCA(uRA4.3, features = VariableFeatures(uRA4.3))
# ElbowPlot(uRA4.3, reduction = "pca", ndims = 50)

# Unsupervised Re-Clustering
uRA4.3 = FindNeighbors(uRA4.3, dims = 1:40)
uRA4.3 = FindClusters(uRA4.3, resolution = 0.9)

# Running UMAP
uRA4.3 = RunUMAP(uRA4.3, dims = 1:40)
# UMAPPlot(uRA4.3, label = TRUE)

# Rename Clusters
Idents(uRA4.3, cells = WhichCells(uRA4.3, idents = "0")) = "3.0"
Idents(uRA4.3, cells = WhichCells(uRA4.3, idents = "1")) = "3.1"
Idents(uRA4.3, cells = WhichCells(uRA4.3, idents = "2")) = "3.2"
```
	

# Differential Gene Expression
```{r}

uRA4.3.marks = subset(FindAllMarkers(uRA4.3, only.pos = FALSE, min.pct = 0.25), 
	subset = p_val_adj < 0.05)
pos.marks = subset(uRA4.3.marks, subset = avg_log2FC > 0.0)
neg.marks = subset(uRA4.3.marks, subset = avg_log2FC < 0.0)

write.xlsx(pos.marks %>% group_by(cluster) %>% top_n(-10, p_val_adj), 
	"reports/uRA4_C3_subs.xlsx", sheetName = "Top10_Up", col.names = TRUE, row.names = FALSE)
write.xlsx(neg.marks %>% group_by(cluster) %>% top_n(-10, p_val_adj), )
	"reports/uRA4_C3_subs.xlsx", sheetName = "Top10_Down", col.names = TRUE, row.names = FALSE, append = TRUE)

saveRDS(uRA4.3, file = "data/uRA4_C3_subs") # Save sub-cluster object
```

# Map Subclusters in Base Data
```{r]
uRA4.1 = uRA4
Idents(uRA4.1, cells = WhichCells(uRA4.3, idents = "3.0")) = "3.0"
Idents(uRA4.1, cells = WhichCells(uRA4.3, idents = "3.1")) = "3.1"
Idents(uRA4.1, cells = WhichCells(uRA4.3, idents = "3.2")) = "3.2"
Idents(uRA4.1) = factor(Idents(uRA4.1), 
	levels = c("0", "1", "2", "3.0", "3.1", "3.2", "4")
```

# Identify Gene Signatures
```{r}
# For DN Tbet+ CD11c+ (subcluster 3.1)
DN_sign = subset(FindMarkers(uRA4.1, ident.1 = "3.1", ident.2 = "3.0", min.pct = 0.25, only.pos = TRUE), 
	subset = p_val_adj <0.05)
write.xlsx(DN_sign, file = "reports/uRA4_3.1_vs_3.0_DEGs.xlsx", sheetName = "DEGs", col.names = TRUE, row.names = FALSE)

saveRDS(uRA4.1, file = "data/uRA4_w_subs") # Save Seurat object

# For CD27+ Tbet+ CD11c+ (subcluster 3.2)
CD27_sign = subset(FindMarkers(uRA4.1, ident.1 = "3.2", ident.2 = "3.0", min.pct = 0.25, only.pos = TRUE), 
	subset = p_val_adj <0.05)
write.xlsx(CD27_sign, file = "reports/uRA4_3.2_vs_3.0_DEGs.xlsx", sheetName = "DEGs", col.names = TRUE, row.names = FALSE)

saveRDS(uRA4.1, file = "data/uRA4_w_subs") # Save Seurat object
```
